File: frontendios/src/services/logbookService.ts
/**
 * Service Logbook pour EdgeCoach iOS
 * Gestion du carnet de bord d'entraînement (nutrition, hydratation, notes, météo, équipement)
 */

import { apiService } from './api';

// Types pour la nutrition
export interface NutritionItem {
  uniqueId: string;
  brand: string;
  name: string;
  type: string;
  calories: number;
  carbs: number;
  caffeine: number;
  timingMinutes?: number;
  timingFormatted?: string; // Format "0h30" pour l'API
  quantity: number;
}

export interface NutritionTotals {
  calories: number;
  carbs: number;
  caffeine: number;
}

export interface NutritionData {
  items: NutritionItem[];
  totals: NutritionTotals;
  timeline?: NutritionItem[];
}

// Types pour l'hydratation
export interface HydrationItem {
  id: string;
  name: string;
  content: string; // Type de contenu: 'water', 'electrolytes', 'isotonic', 'energy', 'bcaa'
  quantity: number;
  volume: number;
}

export interface HydrationData {
  items: HydrationItem[];
  totalVolume: number;
}

// Types pour la météo
export interface WeatherData {
  temperature?: string;
  conditions?: string;
}

// Types pour l'équipement de session
export interface SessionEquipment {
  bikes?: string;
  shoes?: string;
  wetsuits?: string;
  [key: string]: string | undefined;
}

// Type pour les données logbook complètes
export interface LogbookData {
  nutrition: NutritionData;
  hydration: HydrationData;
  notes: string;
  weather: WeatherData;
  equipment: SessionEquipment;
  effortRating?: number;
  perceivedEffort?: string;
}

// Type pour sauvegarder les données logbook
export interface SaveLogbookRequest {
  sessionId: string;
  mongoId: string;
  sessionDate?: string;
  sessionName?: string;
  timestamp?: string;
  logbook: Partial<LogbookData>;
}

// Type pour un document logbook complet
export interface LogbookDocument {
  _id?: string;
  user_id: string;
  session_id: string;
  mongo_id: string;
  session_date?: string;
  session_name?: string;
  timestamp?: string;
  created_at?: string;
  logbook_data: {
    sessionId: string;
    mongoId: string;
    logbook: LogbookData;
  };
}

// Résultats API
export interface LogbookResult {
  success: boolean;
  data?: LogbookDocument;
  error?: string;
}

export interface LogbookListResult {
  success: boolean;
  data?: LogbookDocument[];
  count?: number;
  error?: string;
}

export interface SaveLogbookResult {
  success: boolean;
  message?: string;
  formattedText?: string;
  error?: string;
}

/**
 * Service pour gérer le carnet de bord d'entraînement
 */
class LogbookService {
  /**
   * Récupérer tous les logbooks d'un utilisateur
   */
  async getAllLogbooks(userId: string): Promise<LogbookListResult> {
    try {
      const response = await apiService.get<{
        status: string;
        count: number;
        data: LogbookDocument[];
      }>('/logbook', { user_id: userId });

      return {
        success: true,
        data: response.data,
        count: response.count,
      };
    } catch (error: any) {
      console.error('Error fetching all logbooks:', error);

      // 404 = pas de logbooks, ce n'est pas une erreur
      if (error.response?.status === 404) {
        return {
          success: true,
          data: [],
          count: 0,
        };
      }

      return {
        success: false,
        error: error.userMessage || error.message || 'Erreur lors de la récupération des logbooks',
      };
    }
  }

  /**
   * Récupérer le logbook d'une session spécifique par mongoId
   */
  async getLogbookByMongoId(userId: string, mongoId: string): Promise<LogbookResult> {
    try {
      const response = await apiService.get<{
        status: string;
        count: number;
        data: LogbookDocument[];
      }>('/logbook', { user_id: userId, mongo_id: mongoId });

      if (response.data && response.data.length > 0) {
        return {
          success: true,
          data: response.data[0],
        };
      }

      // Pas de logbook trouvé - c'est normal, pas une erreur
      return {
        success: true,
        data: undefined,
      };
    } catch (error: any) {
      // 404 = pas de logbook trouvé, ce n'est PAS une erreur
      if (error.response?.status === 404) {
        // Silencieusement retourner success avec pas de data
        return {
          success: true,
          data: undefined,
        };
      }

      console.error('Error fetching logbook by mongoId:', error);
      return {
        success: false,
        error: error.userMessage || error.message || 'Erreur lors de la récupération du logbook',
      };
    }
  }

  /**
   * Récupérer le logbook d'une session spécifique par sessionId
   */
  async getLogbookBySessionId(userId: string, sessionId: string): Promise<LogbookResult> {
    try {
      const response = await apiService.get<{
        status: string;
        count: number;
        data: LogbookDocument[];
      }>('/logbook', { user_id: userId, session_id: sessionId });

      if (response.data && response.data.length > 0) {
        return {
          success: true,
          data: response.data[0],
        };
      }

      // Pas de logbook trouvé - c'est normal, pas une erreur
      return {
        success: true,
        data: undefined,
      };
    } catch (error: any) {
      // 404 = pas de logbook trouvé, ce n'est PAS une erreur
      if (error.response?.status === 404) {
        return {
          success: true,
          data: undefined,
        };
      }

      console.error('Error fetching logbook by sessionId:', error);
      return {
        success: false,
        error: error.userMessage || error.message || 'Erreur lors de la récupération du logbook',
      };
    }
  }

  /**
   * Sauvegarder les données du logbook pour une session
   */
  async saveLogbook(userId: string, data: SaveLogbookRequest): Promise<SaveLogbookResult> {
    try {
      const payload = {
        sessionId: data.sessionId,
        mongoId: data.mongoId,
        sessionDate: data.sessionDate,
        sessionName: data.sessionName,
        timestamp: data.timestamp || new Date().toISOString(),
        logbook: data.logbook,
      };

      const response = await apiService.post<{
        status: string;
        message: string;
        formatted_text?: string;
      }>(`/logbook?user_id=${userId}`, payload);

      return {
        success: true,
        message: response.message,
        formattedText: response.formatted_text,
      };
    } catch (error: any) {
      console.error('Error saving logbook:', error);
      return {
        success: false,
        error: error.userMessage || error.message || 'Erreur lors de la sauvegarde du logbook',
      };
    }
  }

  /**
   * Extraire les données logbook d'une session (helper)
   */
  extractLogbookData(session: any): LogbookData | null {
    const rawLogbook = session?.file_datas?.logbook || session?.fileData?.logbook;
    if (!rawLogbook) return null;

    // Si structure imbriquée { logbook: {...} }
    if (rawLogbook.logbook) {
      return rawLogbook.logbook;
    }

    // Si structure directe
    return rawLogbook;
  }

  /**
   * Créer un logbook vide avec des valeurs par défaut
   */
  createEmptyLogbook(): LogbookData {
    return {
      nutrition: {
        items: [],
        totals: { calories: 0, carbs: 0, caffeine: 0 },
        timeline: [],
      },
      hydration: {
        items: [],
        totalVolume: 0,
      },
      notes: '',
      weather: {
        temperature: '',
        conditions: '',
      },
      equipment: {},
    };
  }

  /**
   * Calculer les totaux nutritionnels à partir des items
   */
  calculateNutritionTotals(items: NutritionItem[]): NutritionTotals {
    return items.reduce(
      (totals, item) => ({
        calories: totals.calories + (item.calories * item.quantity),
        carbs: totals.carbs + (item.carbs * item.quantity),
        caffeine: totals.caffeine + (item.caffeine * item.quantity),
      }),
      { calories: 0, carbs: 0, caffeine: 0 }
    );
  }

  /**
   * Calculer le volume total d'hydratation
   */
  calculateHydrationTotal(items: HydrationItem[]): number {
    return items.reduce(
      (total, item) => total + (item.volume * item.quantity),
      0
    );
  }

  /**
   * Formater le timing en heures:minutes
   */
  formatTiming(timingMinutes: number): string {
    const hours = Math.floor(timingMinutes / 60);
    const minutes = timingMinutes % 60;
    return `${hours}h${minutes.toString().padStart(2, '0')}`;
  }
}

export const logbookService = new LogbookService();
export default logbookService;

File: EdgeCoach/Services/LogbookService.swift
/**
 * Service Logbook pour EdgeCoach iOS
 * Gestion du carnet de bord d'entraînement (nutrition, hydratation, notes, météo, équipement)
 */

import Foundation

// MARK: - Nutrition Models

struct NutritionItem: Codable, Identifiable {
    let uniqueId: String
    let brand: String
    let name: String
    let type: String
    let calories: Int
    let carbs: Int
    let caffeine: Int
    var timingMinutes: Int?
    var quantity: Int

    var id: String { uniqueId }
}

struct NutritionTotals: Codable {
    var calories: Int
    var carbs: Int
    var caffeine: Int

    static var zero: NutritionTotals {
        NutritionTotals(calories: 0, carbs: 0, caffeine: 0)
    }
}

struct NutritionData: Codable {
    var items: [NutritionItem]
    var totals: NutritionTotals
    var timeline: [NutritionItem]?

    static var empty: NutritionData {
        NutritionData(items: [], totals: .zero, timeline: nil)
    }
}

// MARK: - Hydration Models

struct HydrationItem: Codable, Identifiable {
    let id: String
    let name: String
    let type: String
    var quantity: Int
    let volume: Int // in ml
}

struct HydrationData: Codable {
    var items: [HydrationItem]
    var totalVolume: Int

    static var empty: HydrationData {
        HydrationData(items: [], totalVolume: 0)
    }
}

// MARK: - Weather & Equipment

struct WeatherData: Codable {
    var temperature: String?
    var conditions: String?

    static var empty: WeatherData {
        WeatherData(temperature: nil, conditions: nil)
    }
}

struct SessionEquipmentData: Codable {
    var bikes: String?
    var shoes: String?
    var wetsuits: String?
}

// MARK: - Logbook Data

struct LogbookData: Codable {
    var nutrition: NutritionData
    var hydration: HydrationData
    var notes: String
    var weather: WeatherData
    var equipment: SessionEquipmentData?
    var effortRating: Int?
    var perceivedEffort: String?
    var intervalAnnotations: [IntervalAnnotation]?

    enum CodingKeys: String, CodingKey {
        case nutrition
        case hydration
        case notes
        case weather
        case equipment
        case effortRating = "effort_rating"
        case perceivedEffort = "perceived_effort"
        case intervalAnnotations = "interval_annotations"
    }

    static var empty: LogbookData {
        LogbookData(
            nutrition: .empty,
            hydration: .empty,
            notes: "",
            weather: .empty,
            equipment: nil,
            effortRating: nil,
            perceivedEffort: nil,
            intervalAnnotations: nil
        )
    }
}

// MARK: - API Response Models

struct LogbookDocument: Codable, Identifiable {
    let _id: String?
    let user_id: String
    let session_id: String
    let mongo_id: String
    let session_date: String?
    let session_name: String?
    let timestamp: String?
    let created_at: String?
    let logbook_data: LogbookDataWrapper?

    var id: String { _id ?? mongo_id }
}

struct LogbookDataWrapper: Codable {
    let sessionId: String?
    let mongoId: String?
    let logbook: LogbookData?
}

struct LogbookListResponse: Codable {
    let status: String
    let count: Int
    let data: [LogbookDocument]
}

struct SaveLogbookResponse: Codable {
    let status: String
    let message: String
    let formatted_text: String?
}

// MARK: - Save Request

struct SaveLogbookRequest: Encodable {
    let sessionId: String
    let mongoId: String
    let sessionDate: String?
    let sessionName: String?
    let timestamp: String
    let logbook: LogbookData
}

// MARK: - Logbook Service

@MainActor
class LogbookService {
    static let shared = LogbookService()
    private let api = APIService.shared

    private init() {}

    // MARK: - Get Logbook

    func getLogbookByMongoId(userId: String, mongoId: String) async throws -> LogbookData? {
        do {
            let response: LogbookListResponse = try await api.get("/logbook", queryParams: [
                "user_id": userId,
                "mongo_id": mongoId
            ])

            if let first = response.data.first, let data = first.logbook_data?.logbook {
                return data
            }
            return nil
        } catch let error as APIError {
            // 404 = no logbook found, not an error
            if case .notFound = error {
                return nil
            }
            if case .httpError(let code, _) = error, code == 404 {
                return nil
            }
            throw error
        }
    }

    func getLogbookBySessionId(userId: String, sessionId: String) async throws -> LogbookData? {
        do {
            let response: LogbookListResponse = try await api.get("/logbook", queryParams: [
                "user_id": userId,
                "session_id": sessionId
            ])

            if let first = response.data.first, let data = first.logbook_data?.logbook {
                return data
            }
            return nil
        } catch let error as APIError {
            if case .notFound = error {
                return nil
            }
            if case .httpError(let code, _) = error, code == 404 {
                return nil
            }
            throw error
        }
    }

    func getAllLogbooks(userId: String) async throws -> [LogbookDocument] {
        do {
            let response: LogbookListResponse = try await api.get("/logbook", queryParams: [
                "user_id": userId
            ])
            return response.data
        } catch let error as APIError {
            if case .notFound = error {
                return []
            }
            if case .httpError(let code, _) = error, code == 404 {
                return []
            }
            throw error
        }
    }

    // MARK: - Save Logbook

    func saveLogbook(userId: String, sessionId: String, mongoId: String, logbook: LogbookData, sessionDate: String? = nil, sessionName: String? = nil) async throws {
        let request = SaveLogbookRequest(
            sessionId: sessionId,
            mongoId: mongoId,
            sessionDate: sessionDate,
            sessionName: sessionName,
            timestamp: ISO8601DateFormatter().string(from: Date()),
            logbook: logbook
        )

        let _: SaveLogbookResponse = try await api.post("/logbook?user_id=\(userId)", body: request)
    }

    // MARK: - Helpers

    func calculateNutritionTotals(_ items: [NutritionItem]) -> NutritionTotals {
        items.reduce(into: NutritionTotals.zero) { totals, item in
            totals.calories += item.calories * item.quantity
            totals.carbs += item.carbs * item.quantity
            totals.caffeine += item.caffeine * item.quantity
        }
    }

    func calculateHydrationTotal(_ items: [HydrationItem]) -> Int {
        items.reduce(0) { total, item in
            total + (item.volume * item.quantity)
        }
    }

    func formatTiming(_ minutes: Int) -> String {
        let hours = minutes / 60
        let mins = minutes % 60
        return "\(hours)h\(String(format: "%02d", mins))"
    }
}
